src/main/scala/is/hail/asm4s/CodeWithRefs.scala

```diff
@@ -0,0 +1,35 @@
+package is.hail.asm4s
+
+import scala.language.implicitConversions
+import scala.language.higherKinds
+import scala.reflect.ClassTag
+
+case class CodeM[T](action: (FunctionBuilder[F, _] forSome { type F[_] }) => T) {
+  def map[U](f: T => U): CodeM[U] =
+    CodeM(fb => f(action(fb)))
+  def flatMap[U](f: T => CodeM[U]): CodeM[U] =
+    CodeM(fb => f(action(fb)).action(fb))
+  def withFilter(p: T => Boolean): CodeM[T] =
+    CodeM(fb => { val t = action(fb); if (p(t)) t else throw new RuntimeException() })
+  def filter(p: T => Boolean): CodeM[T] =
+    withFilter(p)
+
+  /**
+    * The user must ensure that this {@code CodeM} refers to no more arguments
+    * than {@code FunctionBuilder} {@code fb} provides.
+    */
+  def run[F[_], R](fb: FunctionBuilder[F, R])(implicit ev: T <:< Code[R]): F[R] = {
+    fb.result(action(fb.asInstanceOf[FunctionBuilder[F, _] forSome { type F[_] }]))()
```

Comment:
As we discussed, we should kill the higher kinds, remove R from the FunctionBuilder type signature, and make F a concrete type.